internal/cuckoo/cuckoo.go

```diff
@@ -0,0 +1,239 @@
+package cuckoo
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/optable/match/internal/hash"
+)
+
+const (
+	// Nhash is the number of hash function used for cuckoo hash
+	Nhash = 3
+	// Maximum number of reinsertions.
+	// Each reinsertion kicks off 1 egg (item) and replace it
+	// with the item being reinserted, and then reinserts the
+	// kicked off egg
+	ReInsertLimit = 200
+	Factor        = 1.4
+	HashFunc      = hash.Highway
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+// Cuckoo represents a 3-way Cuckoo hash table data structure
+// that contains the items, bucket indices of each item and the 3
+// hash functions. The bucket lookup is a lookup table on items which
+// tells us which item should be in the bucket at that index. Upon
+// construction the items slice has an additional nil value prepended
+// so the index of the Cuckoo.items slice is +1 compared to the index
```

Comment:
Why is that necessary, this +1 to the index?

Reply:
My reasoning is that the +1 to the index is so there is an empty item at the beginning of the list. When the bucket lookup points to that first item (0th) no item has been allocated to it. It's a keeper value to indicate which buckets haven't been filled.